### PR TITLE
feat: node network checker

### PR DESCRIPTION
**Pull request by asmie**
Allow recognize type of network for each node. Currently it allows to check if node is from real ZCash network or from any other.

Classification is based on three factors:
start_height - this MUST match
port
agent string

If either port or agent string would not match node still can be considered as ZCash node as long as start_height matches. The only exception is when agent string identifies node as version 6 (which is other network than ZCash) - that is blocker and even when two other conditions are met, node wouldn't be considered as valid ZCash node.

### CODE REVIEW

File: src/tools/crawler/metrics.rs [hunk 1]
```diff
@@ -61,6 +65,56 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
...
+            port_matches = true;
+        }
+
+        let agent = if let Some(agent) = &nodes[node].user_agent {
+            agent.0.clone()
+        } else {
+            "".to_string()
+        };
+        let zcash_re = Regex::new(r"^/MagicBean:(\d)\.(\d)\.(\d)/$").unwrap();
+        let zebra_re = Regex::new(r"^/Zebra:(\d)\.(\d)\.(\d)").unwrap();
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:81]: this provides more context below
```suggestion
        let zcash_regex = Regex::new(r"^/MagicBean:(\d)\.(\d)\.(\d)/$").unwrap();
        let zebra_regex = Regex::new(r"^/Zebra:(\d)\.(\d)\.(\d)").unwrap();
```

File: src/tools/crawler/metrics.rs [hunk 2]
```diff
@@ -61,6 +65,56 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
...
     }
 
+    let mut node_network_types = Vec::with_capacity(num_good_nodes);
+    for node in &good_nodes {
+        let mut port_matches = false;
+        let mut agent_matches = false;
+        let mut height_matches = false;
+        let mut blocker = false;
+
+        if node.port() == ZCASH_P2P_PORT {
```
Comment (kylegranger) [inline src/tools/crawler/metrics.rs:75]: don't really need mut?
```
let port_matches = node.port() == ZCASH_P2P_PORT;
```

File: src/tools/crawler/metrics.rs [hunk 3]
```diff
@@ -61,6 +65,56 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
...
+        }
+
+        let cap_ze = zebra_re.captures(agent.as_str());
+        if cap_ze.is_some() {
+            // Accept all zebra versions
+            agent_matches = true;
+        }
+
+        let height = nodes[node].start_height.unwrap_or(0);
+        if height > MIN_BLOCK_HEIGHT {
```
Comment (kylegranger) [inline src/tools/crawler/metrics.rs:105]: same here
```
let height_matches = height > MIN_BLOCK_HEIGHT;
```

File: src/tools/crawler/metrics.rs [hunk 4]
```diff
@@ -48,7 +52,7 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
     let mut protocol_versions = HashMap::with_capacity(num_known_nodes);
     let mut user_agents = HashMap::with_capacity(num_known_nodes);
 
-    for (_, node) in nodes {
+    for (_, node) in nodes.clone() {
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:55]: can we use `iter()` here? 🕵️‍♀️

**Review by kylegranger — APPROVED** [review]
the criteria are reasonable, and the logic is sound.

File: src/tools/crawler/metrics.rs [hunk 5]
```diff
@@ -61,6 +65,56 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
...
+        let zebra_re = Regex::new(r"^/Zebra:(\d)\.(\d)\.(\d)").unwrap();
+
+        let cap_zc = zcash_re.captures(agent.as_str());
+        if let Some(cap) = cap_zc {
+            let major = cap.get(1).unwrap().as_str().parse::<u32>().unwrap();
+            if major < 6 {
+                // Accept all zcash versions < 6 (6 is Flux)
+                agent_matches = true;
+            } else if major == 6 {
+                blocker = true; // Block all zcash versions 6 (Flux) even if they are on the right port
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:90]: a general comment regarding this part and the line 105:

if the must-condition is not met (`height is not okay` || `major==6`), we can just call
```rust
node_network_types.push(NetworkType::Unknown);
continue;
```

so the logic on line 111 gets easier to read.
also we will btw remove  those extra variables `height_matches` and `blocker`

File: src/tools/crawler/metrics.rs (hunk 2 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Done.

File: src/tools/crawler/metrics.rs (hunk 3 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Done.

File: src/tools/crawler/network.rs [hunk 6]
```diff
@@ -23,6 +23,8 @@ pub struct KnownNode {
     pub protocol_version: Option<ProtocolVersion>,
     /// The node's user agent.
     pub user_agent: Option<VarStr>,
+    /// The node's best block height.
```
Comment (Rqnsom) [inline src/tools/crawler/network.rs:26]: the best block height? 😃

File: src/tools/crawler/metrics.rs [hunk 7]
```diff
@@ -61,6 +65,56 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
         }
     }
 
+    let mut node_network_types = Vec::with_capacity(num_good_nodes);
+    for node in &good_nodes {
+        let mut port_matches = false;
+        let mut agent_matches = false;
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs]: alternative idea:
```suggestion
        let mut port_ok = false;
        let mut agent_ok = false;
```

**Review by Rqnsom — APPROVED** [review]
lgtm! 👍

File: src/tools/crawler/metrics.rs (hunk 1 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Changed.
`re` is generally widely used across Regex library (and samples) to indicate if something is regular expression but changed if that wasn't clear enough.

File: src/tools/crawler/metrics.rs (hunk 4 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Yes, we can. Then `user_agent` needs to be cloned.
Done.

File: src/tools/crawler/network.rs (hunk 6 above)
Comment (asmie) [reply src/tools/crawler/network.rs]: Probably was thinking about something else 😃  Changed.

File: src/tools/crawler/metrics.rs [hunk 8]
```diff
@@ -61,6 +65,50 @@ pub fn new_network_summary(crawler: &Crawler, graph: &Graph<SocketAddr>) -> Netw
...
+            if major < 6 {
+                // Accept all zcash versions < 6 (6 is Flux)
+                agent_matches = true;
+            } else if major == 6 {
+                blocker = true; // Block all zcash versions 6 (Flux) even if they are on the right port
+            }
+        }
+
+        let cap_ze = zebra_re.captures(agent.as_str());
+        if cap_ze.is_some() {
```
[the remark this answers is not in the document]
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Added.

File: src/tools/crawler/metrics.rs (hunk 1 above)
Comment (Rqnsom) [reply src/tools/crawler/metrics.rs]: >  re is generally widely used across Regex library (and samples) to indicate if something is regular expression but changed if that wasn't clear enough.

aha! I wasn't aware. you could have kept it, I would get used to it if it's commonly used elsewhere

File: src/tools/crawler/metrics.rs [hunk 9]
```diff
@@ -29,6 +36,64 @@ impl NetworkMetrics {
...
+
+        // Look for zebra agent like "/Zebra:1.0.0-rc.4/"
+        let cap_ze = zebra_regex.captures(agent.as_str());
+        if cap_ze.is_some() {
+            // Accept all zebra versions
+            agent_matches = true;
+        }
+
+        let height = nodes[node].start_height.unwrap_or(0);
+        if height < MIN_BLOCK_HEIGHT {
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:80]:
```suggestion
 
        // Check if the height is alright - this is a mandatory check for any zcash node implementation.
        let height = nodes[node].start_height.unwrap_or(0);
        if height < MIN_BLOCK_HEIGHT {
```

File: src/tools/crawler/metrics.rs [hunk 10]
```diff
@@ -29,6 +36,64 @@ impl NetworkMetrics {
...
+            node_network_types.push(NetworkType::Unknown);
+            continue;
+        }
+
+        // Height must match and either port or agent must match to recognize node as zcash node and
+        // there were no conditions that explicitly blocks matching.
```
Comment (Rqnsom) [inline src/tools/crawler/metrics.rs:86]:
```suggestion
        // When a block height is correct, we still need one additional confirmation:
        // In rare cases, the agent or the port won't use a commonly used value. 
```

**Review by Rqnsom — APPROVED** [review]
LGTM, I would just add a few extra comments to clarify

(these are basically copied from the PR description we have now, but once we merge it, it won't be so accessible to future readers)

File: src/tools/crawler/metrics.rs (hunk 5 above)
Comment (Rqnsom) [reply src/tools/crawler/metrics.rs]: > Probably depends on one's preferences 

that's so true!
that's why I asked for some extra comments which provide more context while keeping things simple 😃

File: src/tools/crawler/metrics.rs (hunk 9 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Changed.

File: src/tools/crawler/metrics.rs (hunk 10 above)
Comment (asmie) [reply src/tools/crawler/metrics.rs]: Changed.

**Review by Rqnsom — APPROVED** [review]

**Pull request merged**